REGR: Failed to open HDF-EOS file

From: https://github.com/corteva/rioxarray/pull/606
Related: https://github.com/corteva/rioxarray/issues/544

## Expected behavior and actual behavior.

This works with GDAL 3.5.3:
```
FAILED test/integration/test_integration__io.py::test_open_group_filter[open_rasterio_engine] - rasterio.errors.RasterioIOError: Failed to open HDF-EOS file "/__w/rioxarray/rioxarray/test/test_data/input/MOD09GA.A2008296.h14v17.006.2015181011753.hdf" for swath reading.
FAILED test/integration/test_integration_xarray_plugin.py::test_xarray_open_dataset__drop_variables - rasterio.errors.RasterioIOError: HDF4_EOS:EOS_GRID:/__w/rioxarray/rioxarray/test/test_data/input/MOD09GA.A2008296.h14v17.006.2015181011753.hdf:MODIS_Grid_500m_2D:iobs_res_1: No such file or directory
```

## Steps to reproduce the problem.

Input file: https://github.com/corteva/rioxarray/blob/master/test/test_data/input/MOD09GA.A2008296.h14v17.006.2015181011753.hdf?raw=true

Seems related to opening and reading the file multiple times. The first read works okay.
```python
from osgeo import gdal

path = "MOD09GA.A2008296.h14v17.006.2015181011753.hdf"
layers = []
for iii in range(100):
    ds = gdal.Open(path)
    if not ds:
        print(iii)
        break
    subds = gdal.Open(ds.GetSubDatasets()[0][0])
    if not subds:
        print(iii)
        break
    layers.append(subds)
    subds.GetRasterBand(1).GetNoDataValue()
    del ds
```
Output:
```
ERROR 4: Failed to open HDF-EOS file "MOD09GA.A2008296.h14v17.006.2015181011753.hdf" for swath reading.
32
```

## Operating system

Ubuntu 20.04 64 bit

## GDAL version and provenance

GDAL 3.5.3 & 3.6.0 from docker image.

--- a/autotest/gcore/hdf4_read.py
+++ b/autotest/gcore/hdf4_read.py
@@ -94,6 +94,33 @@ def test_hdf4_read_gr_palette():
 
 
 ###############################################################################
+# Test opening more than 32 simultaneous HDF4_EOS files
+
+
+def test_hdf4_more_than_32_files():
+
+    if (
+        gdal.GetDriverByName("HDF4Image").GetMetadataItem("HDF4_HAS_MAXOPENFILES")
+        != "YES"
+    ):
+        pytest.skip("HDF4_HAS_MAXOPENFILES not set")
+
+    if not gdaltest.download_file(
+        "https://gamma.hdfgroup.org/ftp/pub/outgoing/NASAHDFfiles2/eosweb/hdf4/hdfeos2-swath-wo-dimmaps/AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
+        "AMSR_E_L2_Ocean_B01_200206182340_A.hdf",
+    ):
+        pytest.skip()
+
+    tab = []
+    for i in range(33):
+        ds = gdal.Open(
+            'HDF4_EOS:EOS_SWATH:"tmp/cache/AMSR_E_L2_Ocean_B01_200206182340_A.hdf":Swath1:Low_res_sst'
+        )
+        assert ds, i
+        tab.append(ds)
+
+
+###############################################################################
 # Test HDF4_SDS with single subdataset
 
 

--- a/cmake/modules/packages/FindHDF4.cmake
+++ b/cmake/modules/packages/FindHDF4.cmake
@@ -110,6 +110,7 @@ find_package_handle_standard_args(HDF4
 
 # set output variables
 if(HDF4_FOUND)
+  set(HDF4_INCLUDE_DIRS "${HDF4_INCLUDE_DIR}")
   if(NOT TARGET HDF4::HDF4)
       set(INCR 1)
       set(HDF4_TARGETS)
@@ -124,7 +125,7 @@ if(HDF4_FOUND)
 
       add_library(HDF4::HDF4 INTERFACE IMPORTED)
       set_target_properties(HDF4::HDF4 PROPERTIES
-                            INTERFACE_INCLUDE_DIRECTORIES ${HDF4_INCLUDE_DIR}
+                            INTERFACE_INCLUDE_DIRECTORIES ${HDF4_INCLUDE_DIRS}
                             INTERFACE_LINK_LIBRARIES "${HDF4_TARGETS}")
   endif()
 endif()

--- a/frmts/hdf4/CMakeLists.txt
+++ b/frmts/hdf4/CMakeLists.txt
@@ -23,8 +23,32 @@ endif ()
 gdal_standard_includes(gdal_HDF4)
 target_include_directories(gdal_HDF4 PRIVATE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
                                              $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/hdf-eos>)
+
+
+# HDF4 library newer than 4.2.5 has a SDreset_maxopenfiles/SDget_maxopenfiles interface
+# which allows opening many HDF files simultaneously (the max number of files was previously
+# hardcoded and too low, smth. like 32). Search for it and use if available.
+cmake_push_check_state(RESET)
+set(CMAKE_REQUIRED_LIBRARIES ${HDF4_LIBRARIES})
+set(CMAKE_REQUIRED_INCLUDES ${HDF4_INCLUDE_DIRS})
+# We need to use symbols of both libdf (Hclose()) and libmfhdf (SDget_maxopenfiles())
+# otherwise the linking fails due to libmfhdf not explicitly linking to
+# libdf, but still requiring it...
+check_c_source_compiles(
+"
+    #include <hdf.h>
+    #include <mfhdf.h>
+    intn nCurrMax = 0;
+    intn nSysLimit = 0;
+    int main() { Hclose(0); return SDget_maxopenfiles(&nCurrMax, &nSysLimit); }
+    "
+HDF4_HAS_MAXOPENFILES)
+cmake_pop_check_state()
+
 if (HDF4_HAS_MAXOPENFILES)
   target_compile_definitions(gdal_HDF4 PRIVATE -DHDF4_HAS_MAXOPENFILES)
+else()
+  message(WARNING "SDget_maxopenfiles() not found in HDF library. Strange...")
 endif ()
 
 gdal_target_link_libraries(gdal_HDF4 PRIVATE HDF4::HDF4)

--- a/frmts/hdf4/hdf4dataset.cpp
+++ b/frmts/hdf4/hdf4dataset.cpp
@@ -1370,6 +1370,10 @@ void GDALRegister_HDF4()
 
     GetGDALDriverManager()->RegisterDriver( poDriver );
 
+#ifdef HDF4_HAS_MAXOPENFILES
+    poDriver->SetMetadataItem("HDF4_HAS_MAXOPENFILES", "YES");
+#endif
+
 #ifdef HDF4_PLUGIN
     GDALRegister_HDF4Image();
 #endif

--- a/frmts/hdf4/hdf4imagedataset.cpp
+++ b/frmts/hdf4/hdf4imagedataset.cpp
@@ -4044,6 +4044,9 @@ void GDALRegister_HDF4Image()
         "   <Option name='RANK' type='int' description='Rank of output SDS'/>"
         "</CreationOptionList>" );
 
+#ifdef HDF4_HAS_MAXOPENFILES
+    poDriver->SetMetadataItem("HDF4_HAS_MAXOPENFILES", "YES");
+#endif
     poDriver->pfnOpen = HDF4ImageDataset::Open;
     poDriver->pfnCreate = HDF4ImageDataset::Create;
 

